Sparkle/SPUDownloaderSession.m

```diff
@@ -0,0 +1,181 @@
+            
+            [self.delegate downloaderDidFailWithError:error];
+        } else {
+            NSString *downloadFileName = self.desiredFilename;
+            NSString *downloadFileNameDirectory = [tempDir stringByAppendingPathComponent:downloadFileName];
+            
+            NSError *createError = nil;
+            if (![[NSFileManager defaultManager] createDirectoryAtPath:downloadFileNameDirectory withIntermediateDirectories:NO attributes:nil error:&createError]) {
+                NSError *error = [NSError errorWithDomain:SUSparkleErrorDomain code:SUTemporaryDirectoryError userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Can't make a download file name %@ directory inside temporary directory for the update download at %@.", downloadFileName, downloadFileNameDirectory] }];
+                
+                [self.delegate downloaderDidFailWithError:error];
+            } else {
+                NSString *name = self.download.response.suggestedFilename;
+                if (!name) {
+                    name = location.lastPathComponent; // This likely contains nothing useful to identify the file (e.g. CFNetworkDownload_87LVIz.tmp)
+                }
+                NSString *toPath = [downloadFileNameDirectory stringByAppendingPathComponent:name];
+                NSString *fromPath = location.path; // suppress moveItemAtPath: non-null warning
+                NSError *error = nil;
+                if ([[NSFileManager defaultManager] moveItemAtPath:fromPath toPath:toPath error:&error]) {
+                    self.downloadFilename = toPath;
+                    [self.delegate downloaderDidSetDestinationName:name temporaryDirectory:downloadFileNameDirectory];
+                    [self downloadDidFinish];
+                } else {
+                    [self.delegate downloaderDidFailWithError:error];
+                }
+            }
+        }
+    }
+}
+
+- (void)URLSession:(NSURLSession *)__unused session downloadTask:(NSURLSessionDownloadTask *)__unused downloadTask didWriteData:(int64_t)bytesWritten totalBytesWritten:(int64_t)__unused totalBytesWritten totalBytesExpectedToWrite:(int64_t)totalBytesExpectedToWrite
+{
+    
+    if (self.mode == SPUDownloadModePersistent && totalBytesExpectedToWrite > 0 && !self.receivedExpectedBytes) {
+        self.receivedExpectedBytes = YES;
+        [self.delegate downloaderDidReceiveExpectedContentLength:totalBytesExpectedToWrite];
+    }
+    
+    if (self.mode == SPUDownloadModePersistent && bytesWritten >= 0) {
+        [self.delegate downloaderDidReceiveDataOfLength:(uint64_t)bytesWritten];
+    }
+}
+
+- (void)URLSession:(NSURLSession *)__unused session task:(NSURLSessionTask *)__unused task didCompleteWithError:(NSError *)error
+{
+    self.download = nil;
+    if (error) {
```

Comment:
Why is that? Is this method called without an `error`?

I'm worried that if download fails without notifying the delegate it could make the delegate stuck waiting forever.

Reply:
Yeah, the method is called without an error. See [documentation](https://developer.apple.com/documentation/foundation/nsurlsessiontaskdelegate/1411610-urlsession). `If an error occurred, an error object indicating how the transfer failed, otherwise NULL.` My guess is that this is _always_ called for `NSURLSession`, since a session can be called with multiple tasks, and `URLSession:downloadTask:didFinishDownloadingToURL` is also called so you can track tasks as they complete. A better solution may be to refactor the logic from the task finishing to the session completing so that there aren't any weirdo race conditions or anything. Is there any case where more than one task would be running at once?

IIRC from my testing, sending the `nil` error on to the delegate caused errors (which, granted, should likely be fixed). If I recall, in the `ui-separation-and-xpc` branch, `delegate` was `nil` (race condition?) by the time this was called. I don't have the time this instant to retest things, so I'll take a closer look and report back "soon."